### DISCUSSION TITLE
Conditional matrices

### DISCUSSION THREAD

**Discussion by reikdas** [community/community#25083] [Actions]
<p>For my C++ project, I want to install a particular matrix of compilers for a particular runner, how do I structure it?</p>
<p>This is an excerpt of my YAML file -</p>
<pre><code>jobs:
  build:

    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-16.04, ubuntu-18.04, macOS-latest]
        if: runner.os == 'Linux':
          cpp-version: [gcc-4.8, gcc-5, gcc-6, gcc-7, clang-3.9, clang]
        if: runner.os == 'macOS':
          cpp-version: [xcode9.3, xcode11]

steps:
    - uses: actions/checkout@v1
    - name: Setup compiler
      run: |
         if ["$RUNNER_OS" == "Linux"]; then
              apt install ${{ matrix.cpp-version }}
         elif ["$RUNNER_OS" == "macOS"]; then
              port install ${{ matrix.python-version }}
         else
              echo "$RUNNER_OS not supported"
              exit 1
         fi
      shell: bash
</code></pre>
<p>Is this the right way to do it? I get a parsing error on </p>
<pre><code>if: runner.os == 'Linux':
</code></pre>

**Comment by weide-zhou** [comment]
<p>You cannot put ‘if’ in matrix, ‘<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idif" rel="nofollow noopener">jobs.&lt;job_id&gt;.if</a>’ has same level with ‘<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategy" rel="nofollow noopener">jobs.&lt;job_id&gt;.strategy</a>’.</p>
<p>As an alternative, if you would like to keep using one job, you can create an arbitrary name row to the matrix, then match against in include parameter, code as below, sample <a href="https://github.com/DaanDeMeyer/reproc/blob/master/.github/workflows/main.yml" rel="nofollow noopener">here</a>.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/cb67daa136660814d118379a366f3188f9619383.png" data-download-href="/uploads/short-url/t1pqPmOeFdv2NCRlwgn14wj7o6T.png?dl=1" title="Include.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181088691-34caaee1-0592-4d27-8ba4-7f28cc3cb930.png" alt="Include.png" data-base62-sha1="t1pqPmOeFdv2NCRlwgn14wj7o6T" width="638" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/cb67daa136660814d118379a366f3188f9619383_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Include.png</span><span class="informations">922×722 11 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Or you can use ‘exclude’ paramter to exclude some options, code as below:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/2220b167caa72d41f71781689a26a20f411949f2.png" data-download-href="/uploads/short-url/4RUhgbiefH3Em1t4xjfSci2FmJs.png?dl=1" title="exclude.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181088692-3c8fcc32-e18e-47ed-93a7-76602f5038db.png" alt="exclude.png" data-base62-sha1="4RUhgbiefH3Em1t4xjfSci2FmJs" width="417" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/2220b167caa72d41f71781689a26a20f411949f2_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">exclude.png</span><span class="informations">619×741 10.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Or you can create a new job for macOS.</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181088693-bb76f718-329d-4c40-af09-d363c082b9c0.png" alt="newjob.png" data-base62-sha1="ohg2R8sKwDv3LEvf3ikxONyBdDT" width="493" height="371"></p>

**Comment by wipe2238** [comment, marked as answer]
<p>Alternative: use <a href="https://github.community/t/matrix-include/16721/3">matrix as a list</a>.</p>
<pre><code>jobs:
 build:
  runs-on: ${{ matrix.cfg.os }}
  strategy:
   matrix:
    cfg:
  - { os: ubuntu-16.04, cpp-version: gcc-4.8 }
  - { os: ubuntu-16.04, cpp-version: gcc-5 }
 - { os: ubuntu-16.04, cpp-version: gcc-6 }
- { os: ubuntu-16.04, cpp-version: gcc-7 }
- { os: ubuntu-16.04, cpp-version: clang-3.9 }
  - { os: ubuntu-16.04, cpp-version: clang }
  - { os: ubuntu-18.04, cpp-version: gcc-4.8 }
  - { os: ubuntu-18.04, cpp-version: gcc-5 }
  - { os: ubuntu-18.04, cpp-version: gcc-6 }
  - { os: ubuntu-18.04, cpp-version: gcc-7 }
- { os: ubuntu-18.04, cpp-version: clang-3.9 }
  - { os: ubuntu-18.04, cpp-version: clang }
 - { os: macOS-latest, cpp-version: xcode9.3 }
  - { os: macOS-latest, cpp-version: xcode11 }

- run: |
         if ["$RUNNER_OS" == "Linux"]; then
            sudo apt install ${{ matrix.cfg.cpp-version }}
         fi
shell: bash
</code></pre>
<p>Much easier to mantain matrix in future, and hella easier to read. No need to split job for each OS either.</p>

**Comment by reikdas** [comment]
<p>When using the github online editor, I get the error on cfg stating - “Matrix options must only contain primitive values”.</p>

**Comment by wipe2238** [comment]
<blockquote>
<p><a class="mention" href="/reikdas">@reikdas</a> wrote:</p>
</blockquote>
<blockquote>
<p>“Matrix options must only contain primitive values”.</p>
</blockquote>
<p>Huh, sounds like online editor bug/limitation. I never really used it, but i’m sure runner accepts such syntax without problems. Live example: <a href="https://github.com/rotators/ReDefine/blob/636a6676e48402489c88ea42a4874cdc5b98aa03/.github/workflows/Build.yml#L13-L28" rel="nofollow noopener">workflow</a>, <a href="https://github.com/rotators/ReDefine/commit/636a6676e48402489c88ea42a4874cdc5b98aa03/checks?check_suite_id=336262200" rel="nofollow noopener">check run</a>.</p>

**Comment by lots0logs** [comment]
<blockquote>
<p><a class="mention" href="/weide-zhou">@weide-zhou</a> wrote:</p>
</blockquote>
<blockquote>
<p>You cannot put ‘if’ in matrix, ‘<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idif" rel="nofollow noopener">jobs.&lt;job_id&gt;.if</a>’ has same level with ‘<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategy" rel="nofollow noopener">jobs.&lt;job_id&gt;.strategy</a>’.</p>
</blockquote>
<p>I’ve seen this argument made on several posts but it’s not very good considering runs-on is also on the same level as strategy and using martix variables there is allowed. IJS. Please remove the restrictions as it makes users’ jobs harder.</p>
